Improve support for unusual RSA key sizes

bartonjs opened the pull request:
For compatibility with archived keys, the Windows RSA implementations
should be capable of reading 384-bit RSA keys.

RSACryptoServiceProvider and RSAOpenSsl both support generating
8-bit aligned key sizes, but RSACng supports only 64-bit aligned key sizes.
Nevertheless, RSACng can successfully import keys of sizes outside that
constraint (e.g. 1032-bit).

This change adds support for both "small" (384) and "off alignment" (1032)
sized RSA key importing to ensure that it works across all CoreFx
implementations.

Solve the heart of #6673, but certificate tests can't be added in the same build.
cc: @AtsushiKan, @stephentoub

stephentoub commented on src/Common/src/System/Security/Cryptography/RSACng.cs:
```diff
@@ -55,6 +60,30 @@ protected override byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm)
+                KeySize = newKeySize;
+
+                // And now behavior is restored.
+                s_legalKeySizes = s_creationKeySizes;
```
Isn't there a race condition here?  We're changing the value of a static that's shared by all instances... what if multiple instances each come along and call SetKeySize at the same time, with different newKeySizes?

bartonjs replied:
Wow, yes.  legal wasn't supposed to be static.

stephentoub commented:
@dotnet-bot test this please

AtsushiKan commented:
Rather than construct a custom KeySizes array, it’d be simpler and clearer to set a boolean _skipKeySizeCheck field, and have LegalKeySizes return new[] { new KeySizes(minSize:1, maxSize:16384, skipSize: 1) } when _skipKeySizeCheck is true.

bartonjs commented:
@AtsushiKan Why 16384? If someone really wants to spend hours generating a key bigger than that, who are we to say that it shouldn't be importable?  We just say that we won't create it.

AtsushiKan commented:
Sure, crank it all way up to Int32.MaxValue.

bartonjs commented:
Okay, I'm going with Ati's suggestion; but I was concerned about how slow the LegalKeySizes check is, and so am rewriting that helper method to use mod instead of a loop.  Step size of 1 with a large value is sort of killer with iterating :).

bartonjs commented:
Okay, fixed up RSAOpenSsl, and found a bug in it.  Left it as 3 separate commits:
* Make IsLegalSize faster
* Fixup unexpected keysize support
* Fix the RSAOpenSsl free-before-validate bug (and add a test)

AtsushiKan commented on src/Common/src/System/Security/Cryptography/RSAOpenSsl.cs:
```diff
@@ -58,10 +59,28 @@ public override int KeySize
             }
         }
 
+        private void SetKeySize(int newKeySize)
```
The name "SetKeySize()" glosses over its true purpose. "ForceKeySize()" would better distinguish it from setting the KeySize property in the eyes of the casual reader. Copying over the explanatory comment from the desktop version of this would be good so people know we didn't just do this out of paranoia.

Otherwise, LGTM

stephentoub commented on src/Common/src/System/Security/Cryptography/RSACng.cs:
```diff
@@ -55,6 +66,13 @@ protected override byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm)
+        private void SetKeySize(int newKeySize)
+        {
+            _skipKeySizeCheck = true;
+            KeySize = newKeySize;
```
Can KeySize throw?  I'm wondering if the resetting of _skipKeySizeCheck should be in a finally.

bartonjs replied:
It should be that the only reason it can throw is the key size not being valid (which _skip) is suppressing; but it's a good defensive change.

bartonjs commented:
* Changed SetKeySize to ForceSetKeySize
* Added the essay from desktop to RSACng.
* Added a much smaller version to RSAOpenSsl :smile:.
* Changed ForceSetKeySize to use try { set_Property(); } finally { reset boolean }

and squashed it back in to the "Improve support" commit.

AtsushiKan commented:
LGTM.

bartonjs commented:
@dotnet-bot test this please
